src/tracers.jl

```diff
@@ -117,30 +117,36 @@ $(TYPEDEF)
 
 For a higher-level interface, refer to [`hessian_pattern`](@ref).
 
+The last type parameter `shared` is a `Bool` indicating whether the Hessian should be shared among all intermediate scalar quantities.
+
 ## Fields
 $(TYPEDFIELDS)
 """
-struct HessianTracer{G,H} <: AbstractTracer
+struct HessianTracer{G,H,shared} <: AbstractTracer
```

Comment:
The way the code is currently written also assumes mutability of `G` and `H`. And it assumes they are compatible with `union_product!(::H, ::G, ::G)`. There are plenty of things we cannot enforce statically.

I don't think patterns should be revisited, because they add an extra layer of complexity. But if you do want to give them another go, please wait until I have opened the "array storage" PR (as opposed to "set storage") in order to have concrete examples.

Reply:
Yes, don't get me wrong, we'll merge this without any larger detours.

> please wait until I have opened the "array storage" PR

Sure!